csrc/codebase/mmocr/cuda/connected_component.cu

```diff
@@ -0,0 +1,431 @@
+    RelabelStripsKernel<<<blocks, threads, 0, stream_>>>(d_mask, height_, width_, d_label_);
+  }
+  cudaMemcpyAsync(&n_comp_, d_n_comp_, sizeof(int), cudaMemcpyDefault, stream_);
+  if (h_label) {
+    dim3 threads(32, 4);
+    dim3 blocks(div_up(width_, (int)threads.x), div_up(height_, (int)threads.y));
+    // reuse d_comp_area_, which is also an int buffer
+    thrust::for_each_n(thrust::cuda::par.on(stream_), thrust::counting_iterator<int>(0),
+                       height_ * width_, _decode_label_op{d_label_, d_comp_area_});
+    cudaMemcpyAsync(h_label, d_comp_area_, sizeof(int) * size_, cudaMemcpyDefault, stream_);
+  }
+  cudaStreamSynchronize(stream_);
+  return n_comp_;
+}
+
+void ConnectedComponents::Impl::GetStats(const uint8_t* d_mask, const float* d_score,
+                                         std::vector<float>& scores, std::vector<int>& areas) {
+  cudaMemsetAsync(d_comp_score_, 0, sizeof(float) * size_, stream_);
+  cudaMemsetAsync(d_comp_area_, 0, sizeof(int) * size_, stream_);
+  dim3 threads(32, 4);
+  dim3 blocks(div_up(width_, (int)threads.x), div_up(height_, (int)threads.y));
+  ComputeStatsKernel_v2<<<blocks, threads, 0, stream_>>>(d_mask, d_label_, d_score, height_, width_,
+                                                         d_comp_score_, d_comp_area_);
+  scores.resize(n_comp_);
+  areas.resize(n_comp_);
+  cudaMemcpyAsync(scores.data(), d_comp_score_, sizeof(float) * n_comp_, cudaMemcpyDefault,
+                  stream_);
+  cudaMemcpyAsync(areas.data(), d_comp_area_, sizeof(int) * n_comp_, cudaMemcpyDefault, stream_);
+
+  cudaStreamSynchronize(stream_);
+}
+
+void ConnectedComponents::Impl::GetContours(std::vector<std::vector<cv::Point>>& corners) {
+  cudaMemsetAsync(d_contour_size_, 0, sizeof(int), stream_);
+
+  auto d_contour = reinterpret_cast<int2*>(d_contour_);
+  {
+    dim3 threads(32, 4);
+    dim3 blocks(div_up(width_, (int)threads.x), div_up(height_, (int)threads.y));
+    GetContoursKernel<<<blocks, threads, 0, stream_>>>(d_label_, height_, width_, d_contour,
+                                                       d_contour_size_);
+  }
+
+  int contour_size{};
+  cudaMemcpyAsync(&contour_size, d_contour_size_, sizeof(int), cudaMemcpyDefault, stream_);
+
+  cudaStreamSynchronize(stream_);
+
+  std::vector<int2> index_comp(contour_size);
+  cudaMemcpyAsync(index_comp.data(), d_contour_, sizeof(int2) * contour_size, cudaMemcpyDefault,
+                  stream_);
+
+  cudaStreamSynchronize(stream_);
+
+  corners.resize(n_comp_);
+  for (const auto& p : index_comp) {
+    auto comp = p.y;
+    assert(0 <= comp && comp < n_comp_);
+    corners[comp].emplace_back(p.x % width_, p.x / width_);
+  }
+}
+
+void ConnectedComponents::Impl::Resize(int height, int width) {
+  size_t size = height * width;
+  if (size > capacity_) {
+    if (!capacity_) {
+      capacity_ = size;
+    } else {
+      while (capacity_ < size) {
+        capacity_ *= growth_factor_;
+      }
+    }
+    cudaFree(d_label_);
```

Comment:
Do we have a plan to add some kind of allocator module?

Reply:
We have an device allocator module, but it's not good enough yet.